### ISSUE TITLE
Fuzzer 29. Checkpoint and out-of-memory 2

### ISSUE THREAD

**Issue by Tishj** [duckdb/duckdb#6188]
Original query:
```sql
CREATE TABLE t1 (c0 VARCHAR(32));
INSERT INTO t1 VALUES ('a');
CHECKPOINT;
PRAGMA MEMORY_LIMIT='1MB';
SELECT DISTINCT 'b' FROM t1 ORDER BY c0;
```

Smallest reproducable example:
```sql
statement ok
PRAGMA MEMORY_LIMIT='1.2MB';

# can replace 'first' with 'string_agg' (possibly more)
statement ok
SELECT first ('b'), 'test' group by 2;
```

- The `CHECKPOINT` is not the cause of the issue (can be left out)
- `SELECT DISTINCT` transforms to `first`
`SELECT first ('b') FROM t1 group by c0` is equivalent
`SELECT string_agg ('b') FROM t1 group by c0` also causes the issue (though it manifests in a EXC_BAD_ACCESS instead)
- `ORDER BY` is also not part of the cause

I thought `Initialize` was never called but it does get hit before Destroy, only it also gets hit a **second** time, and that pointer never went through Initialize.
When we increase the memory limit to a point where the query completes, both Initialize and Destroy are hit twice in the correct order.
That makes me think that Initialize is never hit because we threw the OOM error, but that is not properly communicated - so in Destroy it's assumed to have been initialized.

After increasing the memory limit slightly, we get a regular error: OOM
`Out of Memory Error: could not allocate block of 262144 bytes (1869824/2000000 used)`

So it looks like at some point it fails to realize it needs more memory than it has available, or the block gets evicted before it's done being used, and that's why the memory is uninitialized

1. We're doing a grouped aggregation, we hit Combine, where we try to create groups, this hits an OOM exception.
2. We throw an InterruptException to cancel the query
3. We hit the invalid bool load, from `~HashAggregateGroupingGlobalState` -> `~HashAggregateGlobalState`

~~It seems that in step 1 we tried to evict enough blocks to complete our allocation, we couldn't evict enough - but I suspect we did find a block that could be evicted.
Then in step 3 we hit this error because the block of memory we are using was not pinned, and thus evicted by step 1.~~

At the place we error in step1, it seems no block was evicted at all
Actually, throughout the entire execution of the test, no blocks are evicted in EvictBlocks

-----

`vector<data_ptr_t> payload_hds_ptrs;` is used to store the state pointers
in `aggregate_hashtable.cpp:441` we Scatter, which hits the OOM exception
after Scatter is finished, InitializeStates would be called

But in the destructor, DestroyStates calls `GroupedAggregateHashTable::PayloadApply` to get the pointers, and we have no way of knowing that the states weren't initialized

We might need a bool to indicate if states are initialized? Then just don't call DestroyStates if not all states are initialized?
Or should we maybe swizzle the state pointers to indicate initialized or not?

**Comment by Tishj** [comment]
It's clear to me what the problem is, and I have written a fix for it, but many existing tests broke.
So I felt it was necessary to branch this out into it's own issue so that the research does not get lost

**Comment by Tishj** [comment]
I can not reproduce this anymore on main, likely fixed by #8475
